### PR TITLE
feat: Add implementation for CloudResultRepository for Result repository

### PR DESCRIPTION
**Pull request by dejanzele**
## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [ ] tested locally
- [ ] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

File: pkg/cloud/data/testresult.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+//
+//func (r *CloudTestResultRepository) GetTestSuiteMetrics(ctx context.Context, name string, limit, last int) (testkube.ExecutionsMetrics, error) {
+//	var result testkube.ExecutionsMetrics
+//	payload := map[string]interface{}{"name": name, "limit": limit, "last": last}
+//	err := executeHTTP(ctx, r.httpClient, r.apiURL, CommandRequest{Command: CmdTestResultGetTestSuiteMetrics, Payload: payload})
+//	if err != nil {
+//		return result, err
+//	}
+//	return result, nil
+//}
```
Comment (vsukhin) [inline pkg/cloud/data/testresult.go:139]: Remove commented file

File: pkg/cloud/data/testresult_models.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+//
+//type GetTestSuiteMetricsTestResultRequest struct {
+//	Name  string `json:"name"`
+//	Limit int    `json:"limit"`
+//	Last  int    `json:"last"`
+//}
+//
+//type GetTestSuiteMetricsTestResultResponse struct {
+//	Metrics testkube.ExecutionsMetrics `json:"metrics"`
+//}
```
Comment (vsukhin) [inline pkg/cloud/data/testresult_models.go:103]: remove commented file

File: cmd/api-server/main.go [hunk 3]
```diff
@@ -120,6 +128,24 @@ func main() {
 	secretClient, err := secret.NewClient(namespace)
 	ui.ExitOnError("Getting secret client", err)
 
+	// agent
+	apiKey := os.Getenv("TESTKUBE_CLOUD_API_KEY")
+	var grpcClient cloud.TestKubeCloudAPIClient
+	mode := ModeStandalone
+	if os.Getenv("TESTKUBE_CLOUD_API_KEY") != "" {
```
Comment (exu) [inline cmd/api-server/main.go:135]: line 132 you've already got apiKey

File: cmd/api-server/main.go [hunk 4]
```diff
@@ -120,6 +128,24 @@ func main() {
...
+		mode = ModeAgent
+	}
+	if mode == ModeAgent {
+		cloudURL := os.Getenv("TESTKUBE_CLOUD_URL")
```
Comment (exu) [inline cmd/api-server/main.go:139]: maybe it's time to use config - and use `envconfig` like in main for API>

File: internal/app/api/v1/server.go [hunk 5]
```diff
@@ -7,6 +7,11 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/kubeshop/testkube/pkg"
+	"github.com/kubeshop/testkube/pkg/datefilter"
+	result2 "github.com/kubeshop/testkube/pkg/repository/result"
```
Comment (exu) [inline internal/app/api/v1/server.go:12]: result2?

File: internal/app/api/v1/testsuites.go [hunk 6]
```diff
@@ -8,13 +8,14 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/kubeshop/testkube/pkg/datefilter"
+	testresult2 "github.com/kubeshop/testkube/pkg/repository/testresult"
```
Comment (exu) [inline internal/app/api/v1/testsuites.go:12]: testresult2

File: pkg/agent/events_test.go [hunk 7]
```diff
@@ -3,6 +3,8 @@ package agent_test
 import (
 	"context"
 	"fmt"
+	"github.com/kubeshop/testkube/pkg/log"
```
Comment (exu) [inline pkg/agent/events_test.go:6]: again this?

File: pkg/cloud/data/testresult.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
+package data
+
+//type CloudTestResultRepository struct {
```
Comment (exu) [inline pkg/cloud/data/testresult.go:3]: comments

File: pkg/repository/config/mongo_test.go [hunk 9]
```diff
@@ -4,11 +4,11 @@ package config
 
 import (
 	"context"
+	"github.com/kubeshop/testkube/pkg/repository/storage"
```
Comment (exu) [inline pkg/repository/config/mongo_test.go:7]: in go there is order in packages 

core 

local 

external 

why to break it ;)?

File: pkg/cloud/data/command.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	CmdTestResultGetTestSuiteMetrics   Command = "testresult_get_test_suite_metrics"
+)
+
+type CommandRequest struct {
+	Command Command `json:"command"`
+	Payload any     `json:"payload"`
+}
+
+func execute(ctx context.Context, client cloud.TestKubeCloudAPIClient, command Command, payload any, apiKey string) (*cloud.CommandResponse, error) {
+	jsonPayload, _ := json.Marshal(payload)
```
Comment (vsukhin) [inline pkg/cloud/data/command.go:61]: do we need to shadow err there?

File: cmd/api-server/main.go [hunk 11]
```diff
@@ -51,6 +54,11 @@ import (
 	"github.com/kubeshop/testkube/pkg/ui"
 )
 
+const (
+	ModeStandalone = "standalone"
+	ModeAgent      = "agent"
+)
```
Comment (vsukhin) [inline cmd/api-server/main.go:60]: I would move them to package

File: cmd/api-server/main.go [hunk 12]
```diff
@@ -120,6 +128,24 @@ func main() {
...
+	// agent
+	apiKey := os.Getenv("TESTKUBE_CLOUD_API_KEY")
+	var grpcClient cloud.TestKubeCloudAPIClient
+	mode := ModeStandalone
+	if os.Getenv("TESTKUBE_CLOUD_API_KEY") != "" {
+		mode = ModeAgent
+	}
+	if mode == ModeAgent {
+		cloudURL := os.Getenv("TESTKUBE_CLOUD_URL")
+		isInsecure := os.Getenv("TESTKUBE_CLOUD_TLS_INSECURE") == "true"
```
Comment (vsukhin) [inline cmd/api-server/main.go:140]: Should we check for bool value ParseBool?

File: pkg/cloud/data/result.go [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+	response, err := execute(ctx, r.cloudClient, CmdResultGetTestMetrics, req, r.apiKey)
+	if err != nil {
+		return testkube.ExecutionsMetrics{}, err
+	}
+	var commandResponse GetTestMetricsResultResponse
+	if err := json.Unmarshal(response.Response, &commandResponse); err != nil {
+		return testkube.ExecutionsMetrics{}, err
+	}
+	return commandResponse.Metrics, nil
+}
```
Comment (vsukhin) [inline pkg/cloud/data/result.go:253]: Looks like Generics will be useful to deduplicate this code. At least in the future)

File: pkg/cloud/data/result_models.go [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+type GetTestMetricsResultRequest struct {
+	Name  string `json:"name"`
+	Limit int    `json:"limit"`
+	Last  int    `json:"last"`
+}
+
+type GetTestMetricsResultResponse struct {
+	Metrics testkube.ExecutionsMetrics `json:"metrics"`
+}
```
Comment (vsukhin) [inline pkg/cloud/data/result_models.go:152]: might be we can have EmptyRequest, EmptyResponse

**Review by vsukhin — COMMENTED** [review]
some comments

File: pkg/repository/config/mongo_test.go (hunk 9 above)
Comment (dejanzele) [reply pkg/repository/config/mongo_test.go:7]: my intelliJ configuration was wrong :/

**Review by vsukhin — APPROVED** [review]

**Pull request merged**